### PR TITLE
feat: implement Odoo queue-based retry with exponential backoff (odoo_sync_queue)

### PR DESCRIPTION
**Pull request by sagarkhandagre998**
# Pull Requests Requirements

- [x] The PR title includes a brief description of the work done, including the
      Issue number if applicable.
- [ ] The PR includes a video showing the changes for the work done.
- [x] The PR title follows conventional commit label standards.
- [x] The changes confirm to the OpenElis Global x3
      [Styleguide and Design](https://uwdigi.atlassian.net/wiki/spaces/OG/pages/621346838/OpenELIS+Global+Style+Guide)
      documentation.
- [x] The changes include tests or are validated by existing tests.
- [x] I have read and agree to the Contributing
      [Guidelines](https://openelis-global.org/community/get-involved/) of this
      project.

### Problem

When Odoo was temporarily unavailable (network issues, restarts, maintenance), invoice creation silently failed with data permanently lost. No retry, no persistence, no recovery.

### Solution
Implemented a queue-based retry mechanism using a new `odoo_sync_queue` database table.

### How It Works
1. **Invoice creation fails** → instead of silent drop, enqueued to `odoo_sync_queue` with `status=PENDING`
2. **Retry job** runs every 5 minutes, picks up `PENDING` items
3. **Exponential backoff**: retry 1 after 1min → retry 2 after 5min → retry 3 after 15min
4. **Max retries exceeded** → `status=FAILED`, flagged for manual review
5. **Success** → `status=COMPLETED` with audit timestamp

### Config Properties (all have safe defaults)
```/dev/null/application.properties#L1-3
org.openelisglobal.odoo.retry.enabled=true
org.openelisglobal.odoo.retry.queue.enabled=true
org.openelisglobal.odoo.retry.interval.ms=300000
```

### Result
- ✅ Zero data loss during Odoo downtime
- ✅ Auto-recovery without manual intervention
- ✅ Complete audit trail (`created_at`, `completed_at`, `last_error`, `retry_count`)
- ✅ Safe on any environment — all changesets use `onFail="MARK_RAN"`
- ✅ No breaking changes — feature is opt-in via `odoo.enabled` property

## Screenshots

[Add relevant screenshots here if applicable]

## Related Issue
https://github.com/DIGI-UW/OpenELIS-Global-2/issues/2273

[Add a link to the related issue or mention it here if applicable]

## Other

[Add any additional information or notes here]

### CODE REVIEW

File: src/main/java/org/openelisglobal/sample/event/listener/SamplePatientUpdateDataCreatedEventListener.java [hunk 1]
```diff
@@ -14,26 +16,54 @@
...
+        String accessionNumber = updateData != null ? updateData.getAccessionNumber() : "unknown";
 
-            // OdooIntegrationService will check if connection is available
-            // No need to check here since the service handles it gracefully
+        try {
             odooIntegrationService.createInvoice(updateData);
+        } catch (OdooUnavailableException e) {
+            LogEvent.logWarn(this.getClass().getSimpleName(), "handleSamplePatientUpdateDataCreatedEvent",
+                    "Odoo unavailable for sample " + accessionNumber + ". Enqueueing for retry.");
+            enqueueForRetry(accessionNumber);
```
[the remark this answers is not in the document]
Comment (sagarkhandagre998) [reply src/main/java/org/openelisglobal/sample/event/listener/SamplePatientUpdateDataCreatedEventListener.java:53]: We are checking this condition inside enqueueForRetry() method .

**Comment by sagarkhandagre998** [comment]
@mozzy11  @mherman22 Fixed all caught issues .

File: src/main/java/org/openelisglobal/odoo/service/OdooSyncQueueService.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            log.warn("Odoo sync permanently FAILED for accession: {} after {} retries. Manual review required.",
+                    item.getAccessionNumber(), newRetryCount);
+        } else {
+            item.setStatus(SyncStatus.PENDING);
+            int delayMinutes = RETRY_DELAYS_MINUTES[Math.min(newRetryCount - 1, RETRY_DELAYS_MINUTES.length - 1)];
+            Timestamp nextRetry = Timestamp.from(Instant.now().plusSeconds(delayMinutes * 60L));
+            item.setNextRetryTime(nextRetry);
+            log.warn("Odoo sync failed for accession: {}. Retry {} of {} scheduled in {} minutes.",
+                    item.getAccessionNumber(), newRetryCount, item.getMaxRetries(), delayMinutes);
+        }
```
[the remark this answers is not in the document]
Comment (sagarkhandagre998) [reply src/main/java/org/openelisglobal/odoo/service/OdooSyncQueueService.java]: Retry count is starting with 0 , so this will not prevents scheduling the 3rd backoff delay but we 'll make retry count starts with `retryCount=1` so there is no confusion

**Review by pmanko — CHANGES_REQUESTED** [review]
Hey @sagarkhandagre998 — nice work on this! The overall architecture is solid: DB-backed queue, exponential backoff, idempotent enqueue with partial unique index, proper layered structure. Good test coverage too.

I see you've addressed most of the Copilot comments already. One issue I want to flag that I think is still present:

**Retry path produces empty invoices**

`createInvoiceForSample()` builds a `SamplePatientUpdateData` with only the `Sample` object set. But `createInvoiceLines()` iterates over `updateData.getSampleItemsTests()` to build the invoice line items — and that will be `null` on the retry path since it's never populated. The result is that retried invoices would be created in Odoo with zero line items (the `ref` would be correct as `OE-{accessionNumber}`, but no actual charges).

Two possible approaches:
1. **Reconstruct the data**: In `createInvoiceForSample()`, look up the sample's tests via `SampleService`/`TestService` and rebuild the `sampleItemsTests` list before calling `createInvoice()`.
2. **Persist the payload**: Serialize the relevant invoice data into the `payload` TEXT column (which already exists on the entity but is never populated) at enqueue time, then deserialize on retry.

Option 1 is simpler and works well since the test associations are durable in the DB. Option 2 is more robust for edge cases where test associations might change between enqueue and retry.

A couple minor things:
- Entity has `retryCount = 1` as field default but `enqueue()` sets it to 0 — works correctly at runtime but the mismatched default could confuse future readers. Consider aligning them.
- The `payload` column in the entity/Liquibase is dead code right now — either use it (option 2 above) or remove it to avoid confusion.

Happy to help work through the fix if useful!

**Comment by sagarkhandagre998** [comment]
@pmanko Thanks a lot for your review, I'll address this all soon.

File: src/main/java/org/openelisglobal/odoo/valueholder/OdooSyncQueue.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+    @Column(name = "accession_number", nullable = false)
+    private String accessionNumber;
+
+    @Enumerated(EnumType.STRING)
+    @Column(name = "status", nullable = false)
+    private SyncStatus status = SyncStatus.PENDING;
+
+    @Column(name = "retry_count", nullable = false)
+    private int retryCount = 1;
```
[the remark this answers is not in the document]
Comment (sagarkhandagre998) [reply src/main/java/org/openelisglobal/odoo/valueholder/OdooSyncQueue.java]: If we make retryCount  = 0 then we should make retryCount >= maxRetries;

**Comment by sagarkhandagre998** [comment]
Now I have rebuild the sampleItemsTests before calling createInvoice() , Please check it once @pmanko

**Comment by Agaba-derrick** [comment]
@sagarkhandagre998  can we solve these merge conflict issues ?

**Comment by mozzy11** [comment]
Thank you for this contribution, and apologies that it sat unreviewed for so long.

We are doing a cleanup of the PR queue: this PR is now more than 3 months old and has unresolved merge conflicts with develop. The codebase has changed substantially in that time (the frontend moved to Vite and is being migrated to TypeScript, the test infrastructure was rewritten, and several modules — test catalog, analyzers, EQA, storage — were reworked), so a mechanical rebase is unlikely to succeed.

Closing as stale per our queue policy. **If the problem you fixed still exists on current develop, please open a fresh PR against the latest code — we are actively reviewing now and it will get prompt attention.** Thanks again for contributing to OpenELIS!
